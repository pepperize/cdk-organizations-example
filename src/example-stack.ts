import { Account, DependencyChain, Organization, OrganizationalUnit, PolicyType } from "@pepperize/cdk-organizations";
import { VerifySesEmailAddress } from "@seeebiii/ses-verify-identities";
import { SecureRootUser } from "aws-bootstrap-kit";
import { OrganizationTrail } from "aws-bootstrap-kit/lib/organization-trail";
import { Annotations, Aspects, Stack, StackProps, Tags } from "aws-cdk-lib";
import { Construct } from "constructs";

export interface OrganizationStackProps extends StackProps {
  /**
   * Email of the new AWS Account
   */
  email: string;
}

export class ExampleStack extends Stack {
  public constructor(scope: Construct, id: string, props: OrganizationStackProps) {
    super(scope, id, props);

    const { email } = props;

    if (this.region != "us-east-1") {
      Annotations.of(this).addError("AWS Organizations is only in region us-east-1 available.");
    }

    // Create or import an organization
    const organization = new Organization(this, "Organization", {}); // default: FeatureSet.ALL

    // Enabling SSM AWS Service access to be able to register delegated adminstrator
    organization.enableAwsServiceAccess("ssm.amazonaws.com");
    organization.enableAwsServiceAccess("config-multiaccountsetup.amazonaws.com");

    // Enabling service control policies in the organization
    organization.enablePolicyType(PolicyType.SERVICE_CONTROL_POLICY);
    organization.enablePolicyType(PolicyType.BACKUP_POLICY);
    organization.enablePolicyType(PolicyType.TAG_POLICY);
    organization.enablePolicyType(PolicyType.AISERVICES_OPT_OUT_POLICY);

    // Create an organizational unit (OU)
    const organizationalUnit = new OrganizationalUnit(this, "ExampleOU", {
      parent: organization.root,
      organizationalUnitName: "example",
    });

    // Split email
    const accountName = "example";
    const emailPrefix = email.split("@", 2)[0];
    const domain = email.split("@", 2)[1];
    const accountEmail = `${emailPrefix}+${accountName}@${domain}`;

    // Create an account, nested in the example organizational unit
    const account = new Account(this, "Example", {
      accountName: accountName,
      email: accountEmail,
      parent: organizationalUnit,
    });
    const verifySesEmailAddress = new VerifySesEmailAddress(this, "VerifyEmail", {
      emailAddress: accountEmail,
      region: this.region,
    });
    verifySesEmailAddress.node.addDependency(account);

    // Apply bootstrap kit
    const organizationTrail = new OrganizationTrail(this, "OrganizationTrail", {
      OrganizationId: organization.organizationId,
    });
    organizationTrail.node.addDependency(organization);
    const secureRootUser = new SecureRootUser(this, "SecureRootUser", email);
    secureRootUser.node.addDependency(organization);

    // Adding tags
    Tags.of(this).add("ExampleTagKey", "ExampleTagValue", {});

    // Sequentially deploy organization resources
    Aspects.of(this).add(new DependencyChain());
  }
}
