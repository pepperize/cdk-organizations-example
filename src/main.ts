import { App } from "aws-cdk-lib";
import { ExampleStack } from "./example-stack";

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new ExampleStack(app, "Example", { env: devEnv });

app.synth();
