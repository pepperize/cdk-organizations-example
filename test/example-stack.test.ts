import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { ExampleStack } from "../src/example-stack";

describe("OrganizationStack", () => {
  it("Should have snapshot", () => {
    const app = new App();
    const stack = new ExampleStack(app, "test", {});

    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });

  it("Should an organization custom resource", () => {
    const app = new App();
    const stack = new ExampleStack(app, "test", {});

    const template = Template.fromStack(stack);
    template.resourceCountIs("Custom::Organization_Organization", 1);
    template.resourceCountIs("Custom::Organization_Root", 1);
  });
});
