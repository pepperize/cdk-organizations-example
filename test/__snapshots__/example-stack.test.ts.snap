// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrganizationStack Should have snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": Object {
      "DependsOn": Array [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "4a575666d1c2c6412590d2a56f328e040a81ad1ef59aecee31ae9b393d05f659.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Tags": Array [
          Object {
            "Key": "ExampleTagKey",
            "Value": "ExampleTagValue",
          },
        ],
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "ExampleTagKey",
            "Value": "ExampleTagValue",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ExampleCreateAccount88E1A10C": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AccountName": "example",
        "Email": "your_email+example@gmail.com",
        "IamUserAccessToBilling": "ALLOW",
        "ImportOnDuplicate": "true",
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "ExampleOUOrganizationProvider6AC389E4",
            "Id",
          ],
        },
        "RemovalPolicy": "retain",
        "RoleName": "OrganizationAccountAccessRole",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "ExamplecdkorganizationsAccountProviderNestedStackcdkorganizationsAccountProviderNestedStackResourceB245092F",
            "Outputs.testExamplecdkorganizationsAccountProviderframeworkonEvent599E32F2Arn",
          ],
        },
      },
      "Type": "Custom::Organizations_Account",
      "UpdateReplacePolicy": "Delete",
    },
    "ExampleOUOrganizationProvider6AC389E4": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
        "OrganizationRootRootCustomResourceBB74F060",
        "OrganizationRootTagscdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResource88CA0520",
        "OrganizationRootTagsTagResourceCBEA7B2F",
      ],
      "Properties": Object {
        "ImportOnDuplicate": "true",
        "Name": "example",
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "OrganizationRootRootCustomResourceBB74F060",
            "Roots.0.Id",
          ],
        },
        "RemovalPolicy": "retain",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "ExampleOUcdkorganizationsOrganizationalUnitProviderNestedStackcdkorganizationsOrganizationalUnitProviderNestedStackResource93C44F78",
            "Outputs.testExampleOUcdkorganizationsOrganizationalUnitProviderframeworkonEventCBED6C9FArn",
          ],
        },
      },
      "Type": "Custom::Organizations_OrganizationalUnitProvider",
      "UpdateReplacePolicy": "Delete",
    },
    "ExampleOUTagsTagResource1D07DD22": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "ExampleOUOrganizationProvider6AC389E4",
        "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
        "OrganizationRootRootCustomResourceBB74F060",
        "OrganizationRootTagscdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResource88CA0520",
        "OrganizationRootTagsTagResourceCBEA7B2F",
      ],
      "Properties": Object {
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "ExampleOUOrganizationProvider6AC389E4",
            "Id",
          ],
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "ExampleOUTagscdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResourceD24A6505",
            "Outputs.testExampleOUTagscdkorganizationsTagResourceProviderframeworkonEvent82979B67Arn",
          ],
        },
        "Tags": Array [
          Object {
            "Key": "ExampleTagKey",
            "Value": "ExampleTagValue",
          },
        ],
      },
      "Type": "Custom::Organizations_TagResource",
      "UpdateReplacePolicy": "Delete",
    },
    "ExampleOUTagscdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResourceD24A6505": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "ExampleOUOrganizationProvider6AC389E4",
        "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
        "OrganizationRootRootCustomResourceBB74F060",
        "OrganizationRootTagscdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResource88CA0520",
        "OrganizationRootTagsTagResourceCBEA7B2F",
      ],
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "ExampleTagKey",
            "Value": "ExampleTagValue",
          },
        ],
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.",
              Object {
                "Ref": "AWS::Region",
              },
              ".",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
              },
              "/f3b6b56596c585c4567789956d9bf22ba56b7844eac886b9b1a000067ae1d21e.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "ExampleOUcdkorganizationsOrganizationalUnitProviderNestedStackcdkorganizationsOrganizationalUnitProviderNestedStackResource93C44F78": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
        "OrganizationRootRootCustomResourceBB74F060",
        "OrganizationRootTagscdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResource88CA0520",
        "OrganizationRootTagsTagResourceCBEA7B2F",
      ],
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "ExampleTagKey",
            "Value": "ExampleTagValue",
          },
        ],
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.",
              Object {
                "Ref": "AWS::Region",
              },
              ".",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
              },
              "/2d590d74b01d7eed63962f450229fb759d99ee7d0f61b8699ee502807eba6b15.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "ExampleTagsTagResource07FC6C8C": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "ExampleCreateAccount88E1A10C",
      ],
      "Properties": Object {
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "ExampleCreateAccount88E1A10C",
            "AccountId",
          ],
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "ExampleTagscdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResource342D4093",
            "Outputs.testExampleTagscdkorganizationsTagResourceProviderframeworkonEventE30F8A12Arn",
          ],
        },
        "Tags": Array [
          Object {
            "Key": "ExampleTagKey",
            "Value": "ExampleTagValue",
          },
        ],
      },
      "Type": "Custom::Organizations_TagResource",
      "UpdateReplacePolicy": "Delete",
    },
    "ExampleTagscdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResource342D4093": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "ExampleCreateAccount88E1A10C",
      ],
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "ExampleTagKey",
            "Value": "ExampleTagValue",
          },
        ],
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.",
              Object {
                "Ref": "AWS::Region",
              },
              ".",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
              },
              "/26ccf7eee36f01bd68c0da8038c1265a18f8400db21af4ed1b1afa20e5a1faa3.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "ExamplecdkorganizationsAccountProviderNestedStackcdkorganizationsAccountProviderNestedStackResourceB245092F": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "ExampleTagKey",
            "Value": "ExampleTagValue",
          },
        ],
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.",
              Object {
                "Ref": "AWS::Region",
              },
              ".",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
              },
              "/cd17767e174951859fc0b53e60d9fb3ce04ab88ab3c0a14cf1f8e532ecfa758d.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "Organization06E16095": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "FeatureSet": "ALL",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "OrganizationcdkorganizationsOrganizationProviderNestedStackcdkorganizationsOrganizationProviderNestedStackResource87A8ECA1",
            "Outputs.testOrganizationcdkorganizationsOrganizationProviderframeworkonEventE6788265Arn",
          ],
        },
      },
      "Type": "Custom::Organizations_Organization",
      "UpdateReplacePolicy": "Delete",
    },
    "OrganizationRootRootCustomResourceBB74F060": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "Organization06E16095",
        "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
      ],
      "Properties": Object {
        "Create": "{\\"service\\":\\"Organizations\\",\\"action\\":\\"listRoots\\",\\"region\\":\\"us-east-1\\",\\"physicalResourceId\\":{\\"responsePath\\":\\"Roots.0.Id\\"}}",
        "Delete": "{\\"service\\":\\"Organizations\\",\\"action\\":\\"listRoots\\",\\"region\\":\\"us-east-1\\"}",
        "InstallLatestAwsSdk": false,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": "{\\"service\\":\\"Organizations\\",\\"action\\":\\"listRoots\\",\\"region\\":\\"us-east-1\\",\\"physicalResourceId\\":{\\"responsePath\\":\\"Roots.0.Id\\"}}",
      },
      "Type": "Custom::Organizations_Root",
      "UpdateReplacePolicy": "Delete",
    },
    "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E": Object {
      "DependsOn": Array [
        "Organization06E16095",
      ],
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "organizations:ListRoots",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "OrganizationRootTagsTagResourceCBEA7B2F": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "Organization06E16095",
        "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
        "OrganizationRootRootCustomResourceBB74F060",
      ],
      "Properties": Object {
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "OrganizationRootRootCustomResourceBB74F060",
            "Roots.0.Id",
          ],
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "OrganizationRootTagscdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResource88CA0520",
            "Outputs.testOrganizationRootTagscdkorganizationsTagResourceProviderframeworkonEventE8008E3EArn",
          ],
        },
        "Tags": Array [
          Object {
            "Key": "ExampleTagKey",
            "Value": "ExampleTagValue",
          },
        ],
      },
      "Type": "Custom::Organizations_TagResource",
      "UpdateReplacePolicy": "Delete",
    },
    "OrganizationRootTagscdkorganizationsTagResourceProviderNestedStackcdkorganizationsTagResourceProviderNestedStackResource88CA0520": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "Organization06E16095",
        "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
        "OrganizationRootRootCustomResourceBB74F060",
      ],
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "ExampleTagKey",
            "Value": "ExampleTagValue",
          },
        ],
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.",
              Object {
                "Ref": "AWS::Region",
              },
              ".",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
              },
              "/8bcdbe9b2d8d8dd5f19242a16cc1a3d1db665cebc04250a833b03c9a76814c3a.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "OrganizationcdkorganizationsOrganizationProviderNestedStackcdkorganizationsOrganizationProviderNestedStackResource87A8ECA1": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "ExampleTagKey",
            "Value": "ExampleTagValue",
          },
        ],
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.",
              Object {
                "Ref": "AWS::Region",
              },
              ".",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
              },
              "/0cb00b1c4fc2daf6e8807df09d23c8fcfe89174c3722ee32bac703d7d5b74488.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
